GSW/testUtility/testNS2.pyw

```diff
@@ -26,87 +26,148 @@ commandField = Entry(commandFrame, width=30, borderwidth=3, font=('Consolas',10)
 portField = Entry(serialFrame, width=30, borderwidth=3)
 portField.insert(0, 'COM3')
 
 
 # ==== define functions and buttons ====
-# update log
-def updateLog(text):
+# update log #
+def updateLog(text): # writes text to the log
     log.configure(state ='normal')
-    log.insert(INSERT, text + '\n')
+    log.insert(END, text + '\r\n')
     log.configure(state ='disabled')
     log.see(END)
 
-# send command
-def sendCommand(event=None):
+# send command #
+def sendCommand(event=None): # sends a command over serial
+    if not portOpen:
+        updateLog('No connection!')
+        return
     command = commandField.get()
     if command:
         teensy.write(bytes(command, 'utf-8'))
         updateLog('Command Sent: ' + command)
         commandField.delete(0, END)
-button_sendCommand = Button(commandFrame, text='Send', command=sendCommand)
-# bind the enter key to send command
-root.bind('<Return>', sendCommand)
 
-# read serial
-def readSerial():
+# read serial #
+def readSerial(): # reads incoming communication from NS2 via the open serial port
     data = str(teensy.readline())[2:-1]
     if data:
         monitor.configure(state ='normal')
-        monitor.insert(INSERT, codecs.decode(data, "unicode_escape"))
+        monitor.insert(END, codecs.decode(data, "unicode_escape")) # insert data in monitor
         monitor.configure(state ='disabled')
-        monitor.see(END)
+        if autoScroll.get():
+            monitor.see(END) 
 
 # scan for serial ports
-def scanPorts():
+def scanPorts(): # finds and lists all available com ports
     availablePorts = list_ports.comports()
     updateLog('==== Found Serial Ports: ====')
     for port in availablePorts:
         updateLog('- ' + port.description)
-button_scanPorts = Button(serialFrame, text='Scan For Serial Ports', command=scanPorts)
-        
-# update serial button
-def updateSerialPort():
-    teensy.port = portField.get()
-button_updatePort = Button(serialFrame, text='Set Serial Port', command=updateSerialPort)
+                
+# open serial port #
+def openSerialPort(): # opens a new serial connection if port exists
+    availablePorts = list_ports.comports()
+    for port in availablePorts: # if port exists, connect to it
+        if portField.get() in port.description:
+            teensy.close()
+            teensy.port = portField.get()
+            teensy.open()
+            global portOpen
+            portOpen = True
+            updateLog('Connected to port \"' + portField.get() + '\".')
+            return    
+    updateLog('Port \"' + portField.get() + '\" not found.')
+    
+# save to file #
+def saveFile(): # saves monitor contents to a file
+    fileName = fileNameField.get() + '.txt'
+    filePath = 'saved files\\' + fileNameField.get()
+    if (os.path.exists(filePath)): # display warning if file already exists
+        warning = '\"' + fileName + '\" already exists. Do you wish to overwrite it?'
```

Comment:
"... already exists in \"saved files\" folder. Do you ...."

Reply:
Done